File: nodeJs/tic-tac-toe-board-game/testCase/resultAnalyzeTest.js
let assert = require('assert');
let expect = require('chai').expect;
let ResultAnalyze = require('../resultAnalyze');
let Board = require('../board');
let mark = require('../mark');

describe('mocha resultAnalyze test', () => {
    it('should throw an nullPointerException if you create object of resultAnalyze in pass null', () => {
        expect(() => new ResultAnalyze(null)).to.throw();
    });
    it('should result is win if three same mark in row', () => {
        let board = new Board();
        board.setBoardCellMark(0, mark.O);
        board.setBoardCellMark(1, mark.O);
        board.setBoardCellMark(2, mark.O);
        let resultAnalyze = new ResultAnalyze(board);

        assert.equal(resultAnalyze.genrateResultState(), 'Win');
    });
    it('should result is win if three same mark in column', () => {
        let board = new Board();
        board.setBoardCellMark(0, mark.O);
        board.setBoardCellMark(3, mark.O);
        board.setBoardCellMark(6, mark.O);
        let resultAnalyze = new ResultAnalyze(board);

        assert.equal(resultAnalyze.genrateResultState(), 'Win');
    });
    it('should result is draw if board all cell is fill', () => {
        let board = new Board();
        board.setBoardCellMark(0, mark.O);
        board.setBoardCellMark(1, mark.X);
        board.setBoardCellMark(2, mark.O);
        board.setBoardCellMark(3, mark.O);
        board.setBoardCellMark(4, mark.O);
        board.setBoardCellMark(5, mark.X);
        board.setBoardCellMark(6, mark.X);
        board.setBoardCellMark(7, mark.O);
        board.setBoardCellMark(8, mark.X);
        let resultAnalyze = new ResultAnalyze(board);

        assert.equal(resultAnalyze.genrateResultState(), 'Draw');
    });
    it('should result is in_progress if board all cell is not fill', () => {
        let board = new Board();
        board.setBoardCellMark(0, mark.O);
        board.setBoardCellMark(1, mark.X);
        board.setBoardCellMark(2, mark.O);
        let resultAnalyze = new ResultAnalyze(board);

        assert.equal(resultAnalyze.genrateResultState(), 'In_Progress');
    });

    it('should wining sign is X if X sign is Win', () => {
        let board = new Board();
        board.setBoardCellMark(0, mark.X);
        board.setBoardCellMark(1, mark.X);
        board.setBoardCellMark(2, mark.X);
        let resultAnalyze = new ResultAnalyze(board);

        resultAnalyze.genrateResultState();
        assert.equal(resultAnalyze.getWiningPlayerSign(), mark.X);
    });
});

File: OOP/tic-tac-toe-board-game-app/src/com/techlab/test/cases/ResultAnalyzeTest.java
package com.techlab.test.cases;

import static org.junit.Assert.*;

import org.junit.Test;

import com.techlab.Exception.AllReadyMarkCellException;
import com.techlab.board.Board;
import com.techlab.board.Mark;
import com.techlab.result.analyze.ResultAnalyze;

public class ResultAnalyzeTest {
	private ResultAnalyze resultAnalyze;

	@Test(expected = NullPointerException.class)
	public void testResultAnalyzeConstractorPassingValueNull_PassBoardObject_ThrowNullPointerException() {
		resultAnalyze = new ResultAnalyze(null);
	}

	@Test
	public void testResultAnalyzeConstractorPassingValueNotNull_PassBoardObject_ShouldBeNotThrowNullPointerException() {
		resultAnalyze = new ResultAnalyze(new Board());
	}

	@Test
	public void testGenrateResultState_ThreeSameMarkInRow_ResultGetResultMatchWin() throws AllReadyMarkCellException {

		Board board = new Board();
		board.setBoardCellMark(0, Mark.O);
		board.setBoardCellMark(1, Mark.O);
		board.setBoardCellMark(2, Mark.O);
		ResultAnalyze resultAnalyze = new ResultAnalyze(board);

		String actual = resultAnalyze.genrateResultState().toString();
		String expected = "Win";
		assertEquals(expected, actual);

	}

	@Test
	public void testGenrateResultState_ThreeSameMarkInColumn_ResultGetResultMatchWin()
			throws AllReadyMarkCellException {

		Board board = new Board();
		board.setBoardCellMark(0, Mark.O);
		board.setBoardCellMark(3, Mark.O);
		board.setBoardCellMark(6, Mark.O);
		ResultAnalyze resultAnalyze = new ResultAnalyze(board);

		String actual = resultAnalyze.genrateResultState().toString();
		String expected = "Win";
		assertEquals(expected, actual);

	}

	@Test
	public void testGenrateResultState_BoardAllCellIsFill_ResultGetResultMatchDraw() throws AllReadyMarkCellException {

		Board board = new Board();
		board.setBoardCellMark(0, Mark.O);
		board.setBoardCellMark(1, Mark.X);
		board.setBoardCellMark(2, Mark.O);
		board.setBoardCellMark(3, Mark.O);
		board.setBoardCellMark(4, Mark.O);
		board.setBoardCellMark(5, Mark.X);
		board.setBoardCellMark(6, Mark.X);
		board.setBoardCellMark(7, Mark.O);
		board.setBoardCellMark(8, Mark.X);
		ResultAnalyze resultAnalyze = new ResultAnalyze(board);

		String actual = resultAnalyze.genrateResultState().toString();
		String expected = "Draw";
		assertEquals(expected, actual);

	}

	@Test
	public void testGenrateResultState_BoardAllCellIsNotFill_ResultGetResultMatchInProgress()
			throws AllReadyMarkCellException {

		Board board = new Board();
		board.setBoardCellMark(0, Mark.O);
		board.setBoardCellMark(1, Mark.X);
		board.setBoardCellMark(2, Mark.O);
		ResultAnalyze resultAnalyze = new ResultAnalyze(board);

		String actual = resultAnalyze.genrateResultState().toString();
		String expected = "In_Progress";
		assertEquals(expected, actual);

	}

	@Test
	public void testGetWiningPlayerSign_BoardAllCellInPutTheMark_ResultGetResultWiningPlayerSignO()
			throws AllReadyMarkCellException {

		Board board = new Board();
		board.setBoardCellMark(0, Mark.X);
		board.setBoardCellMark(1, Mark.X);
		board.setBoardCellMark(2, Mark.X);
		board.setBoardCellMark(3, Mark.X);
		board.setBoardCellMark(4, Mark.X);
		board.setBoardCellMark(5, Mark.O);
		board.setBoardCellMark(6, Mark.O);
		board.setBoardCellMark(7, Mark.X);
		board.setBoardCellMark(8, Mark.X);
		ResultAnalyze resultAnalyze = new ResultAnalyze(board);

		resultAnalyze.genrateResultState();
		String actual = resultAnalyze.getWiningPlayerSign();
		String expected = "X";

		assertEquals(expected, actual);
	}

}
